### PR TITLE
comby: init at 1.5.1

### PR DESCRIPTION
**Pull request by sephii**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sephii** [comment]
Note that I didn’t put anything in the maintainers list because I’m not sure I can maintain it since I don’t know anything about the ocaml ecosystem (I just wanted this tool to be packaged). ^^

I also had to upgrade some dependencies. Let me know if you prefer separate pull requests for these, I wasn’t sure how to do it since this PR depends on these new versions.

Ping @sternenseemann, if you’re still interested in reviewing it.

File: pkgs/development/ocaml-modules/mparser-pcre/default.nix
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/mparser-pcre/default.nix:1]: `ocamlPackages.callPackage` is scoped:

```suggestion
{ lib, fetchFromGitHub, buildDunePackage, ocaml_pcre, mparser }:
```

File: pkgs/development/ocaml-modules/mparser-pcre/default.nix
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/mparser-pcre/default.nix:15]:
```suggestion
  buildInputs = [ ocaml_pcre mparser ];
```

File: pkgs/development/ocaml-modules/mparser/default.nix
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/mparser/default.nix:1]:
```suggestion
{ lib, fetchFromGitHub, buildDunePackage }:
```

File: pkgs/development/ocaml-modules/mparser/default.nix
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/mparser/default.nix:20]: Unnecessary for `buildDunePackage`.

```suggestion
```

File: pkgs/development/ocaml-modules/mparser-pcre/default.nix
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/mparser-pcre/default.nix:23]: Simplify by taking a view values from `mparser`:

```suggestion
  meta = mparser.meta // {
    description = "PCRE-based regular expressions";
  };
```

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:8]: According to opam.

```suggestion
      minimumOcamlVersion = "4.08.1";
```

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:1]: Take advantage of `callPackage`.

```suggestion
{  ocamlPackages, fetchFromGitHub, lib, zlib, pkg-config }:
```

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:19]:
```suggestion
        zlib
        pkg-config
```

File: pkgs/top-level/all-packages.nix
Comment (sternenseemann) [inline pkgs/top-level/all-packages.nix:10319]: This should get its own subdirectory, i. e. `../development/toosl/comby/default.nix`.

File: pkgs/top-level/ocaml-packages.nix
Comment (sternenseemann) [inline pkgs/top-level/ocaml-packages.nix:765]: Consider moving `mparser-pcre` to `../development/ocaml-modules/mparser/pcre.nix`. I've started to establish this convention for multiple ocaml packages which are built from the same source archive and share a common prefix.

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:50]: Consider adding an `extraBuildInputs` parameter rather than overriding the derivation here.

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:44]:
```suggestion
        homepage = "https://comby.dev";
```

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:19]: `pkg-config` belongs in `nativeBuildInputs`.

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:26]: The `ppx_*` packages and `dune-configurator` are also `nativeBuildInputs`.

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:10319]:
```suggestion
  comby = callPackage ../development/tools/comby { };
```

File: pkgs/development/ocaml-modules/mparser/pcre.nix
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/mparser/pcre.nix:13]: You can just reuse the source from `mparser`, makes updating easier as well:

```suggestion
  inherit (mparser) src;
```

File: pkgs/development/ocaml-modules/mparser/pcre.nix
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/mparser/pcre.nix:5]: You can also reuse the version attr from mparser.

```suggestion
  inherit (mparser) version;
```

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:27]: We should further reduce the build inputs `mkCombyPackage` passes by default — comby-kernel only uses a fraction of the (native) build inputs comby uses.

File: pkgs/development/ocaml-modules/parany/default.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/parany/default.nix:5]: Please set `minimumOCamlVersion = "4.03"`.

File: pkgs/development/tools/comby/default.nix
Comment (sephii) [reply pkgs/development/tools/comby/default.nix:24]: I’m not sure how to find out which packages a package depends on. I checked the `dune` file in every subpackage, but maybe there’s a more efficient way of doing this? Anyway, I gave it a shot, could you check?

Comment (sternenseemann) [reply pkgs/development/tools/comby/default.nix:24]: I usually check the ones it lists in its `.opam` file.

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:41]: neither `ppx_here` nor `re` are listed in the opam file, are they really necessary (sometimes there are discrepancies though)?

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:29]: `angstrom` is a common dependency according to the opam files.

File: pkgs/development/tools/comby/default.nix
Comment (sternenseemann) [inline pkgs/development/tools/comby/default.nix:58]: opam file also lists:

* conduit-lwt-unix
* hack_parallel
* lwt_react
* tls

Maybe there are optional dependencies as well? It lists the system dependencies gmp, libev as well…

**Comment by sternenseemann** [comment]
Have you tried running the tests for comby-kernel?

File: pkgs/development/tools/comby/default.nix
Comment (sephii) [reply pkgs/development/tools/comby/default.nix]: No they’re not, I removed them.

File: pkgs/development/tools/comby/default.nix
Comment (sephii) [reply pkgs/development/tools/comby/default.nix:32]: Moved to the common dependencies.

File: pkgs/development/tools/comby/default.nix
Comment (sephii) [reply pkgs/development/tools/comby/default.nix:66]: I’ve added the dependencies to the inputs. Maybe some of them are runtime dependencies and the program would crash under certain circumstances if they’re not included?

**Comment by sephii** [comment]
I added `doCheck = true;` (and had to add `cacert` to the check inputs), but didn’t see any test output, so I’m not 100% sure the tests run correctly.

I’ve also noticed that in the meantime, Comby 1.6.0 was released. I’d be happy to update this PR if I can get it to build. Right now I’m getting build errors because `comby-kernel` needs `comby.vangstrom` (as stated [in the dune file](https://github.com/comby-tools/comby/blob/master/lib/kernel/matchers/dune)), but I don’t know how to package `comby.vangstrom` (it seems to be in the `lib/vendored` directory).

**Comment by kevincox** [comment]
I just tried this an it works well. I would love to get this in. It seems as though there are no blockers. It would be nice to merge the current version and then we can update later. Any objections?

**Comment by kevincox** [comment]
Result of `nixpkgs-review pr 123986` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>comby</li>
  </ul>
</details>

**Comment by sternenseemann** [comment]
Sounds alright, conflict needs to be resolved and I was meaning to make another pass reviewing.

**Review by kevincox — APPROVED** [review]

**Comment by sephii** [comment]
I’ve rebased on master and fixed the conflict.

**Comment by kevincox** [comment]
Result of `nixpkgs-review pr 123986` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>comby</li>
  </ul>
</details>

**Review by kevincox — APPROVED** [review]
Looks good to me. I'll give @sternenseemann a day or two to look over again before merging.

File: pkgs/development/ocaml-modules/mparser/pcre.nix
Comment (vbgl) [inline pkgs/development/ocaml-modules/mparser/pcre.nix:9]:
```suggestion
  propagatedBuildInputs = [ ocaml_pcre mparser ];
```

**Comment by kevincox** [comment]
@sephii do you have a chance to address @vbgl's comment and we can merge?

**Comment by sephii** [comment]
@kevincox here you go, sorry for the delay. I’ve also rebased on master.

**Comment by kevincox** [comment]
No problem. Thanks for the contribution!

**Comment by happysalada** [comment]
Some people might be interested in this PR https://github.com/NixOS/nixpkgs/pull/139408

**Pull request merged**
